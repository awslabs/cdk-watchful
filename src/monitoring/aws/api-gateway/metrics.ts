import { Metric, Statistic } from '@aws-cdk/aws-cloudwatch';
import { Duration } from '@aws-cdk/core';
import { WatchedOperation } from '../../../api-gateway';

const enum Metrics {
  FourHundredError = '4XXError',
  FiveHundredError = '5XXError',
  CacheHitCount = 'CacheHitCount',
  CacheMissCount = 'CacheMissCount',
  Count = 'Count',
  IntegrationLatency = 'IntegrationLatency',
  Latency = 'Latency',
}

const Namespace = 'AWS/ApiGateway';

export class ApiGatewayMetricFactory {
  metricErrors(apiName: string, stage: string, op?: WatchedOperation) {
    return {
      errors4XX: this.metric(Metrics.FourHundredError, apiName, stage, op).with({ label: '4XX Errors', statistic: Statistic.SUM }),
      errors5XX: this.metric(Metrics.FiveHundredError, apiName, stage, op).with({ label: '5XX Errors', statistic: Statistic.SUM }),
    };
  }

  metricCache(apiName: string, stage: string, op?: WatchedOperation) {
    return {
      hits: this.metric(Metrics.CacheHitCount, apiName, stage, op).with({ label: 'Cache Hit', statistic: Statistic.SUM }),
      misses: this.metric(Metrics.CacheMissCount, apiName, stage, op).with({ label: 'Cache Miss', statistic: Statistic.SUM }),
    };
  }

  metricCalls(apiName: string, stage: string, op?: WatchedOperation) {
    return this.metric(Metrics.Count, apiName, stage, op).with({ label: 'Calls', statistic: Statistic.SUM });
  }

  metricIntegrationLatency(apiName: string, stage: string, op?: WatchedOperation) {
    return {
      min: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'min', statistic: Statistic.MINIMUM }),
      avg: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'avg', statistic: Statistic.AVERAGE }),
      p90: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p90', statistic: 'p90' }),
      p95: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p95', statistic: 'p95' }),
      p99: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p99', statistic: 'p99' }),
      max: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'max', statistic: Statistic.MAXIMUM }),
    };
  }

  metricLatency(apiName: string, stage: string, op?: WatchedOperation) {
    return {
      min: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'min', statistic: Statistic.MINIMUM }),
      avg: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'avg', statistic: Statistic.AVERAGE }),
      p90: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p90', statistic: 'p90' }),
      p95: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p95', statistic: 'p95' }),
      p99: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p99', statistic: 'p99' }),
      max: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'max', statistic: Statistic.MAXIMUM }),
    };
  }

  protected metric(metricName: Metrics, apiName: string, stage: string, op?: WatchedOperation) {
    return new Metric({
      metricName,
      namespace: Namespace,
      period: Duration.minutes(1),
      dimensions: {
        ApiName: apiName,
        Stage: stage,
        ...op && {
          Method: op.httpMethod,
          Resource: op.resourcePath,
        },
      },
    });
  }
}

