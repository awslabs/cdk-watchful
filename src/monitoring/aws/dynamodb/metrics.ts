import { Metric } from '@aws-cdk/aws-cloudwatch';
import { Duration } from '@aws-cdk/core';

export class DynamoDbMetricFactory {
  static readonly Namespace = 'AWS/DynamoDB';

  metricConsumedCapacityUnits(tableName: string) {
    function metric(metricName: string, label: string) {
      return new Metric({
        metricName,
        label,
        namespace: DynamoDbMetricFactory.Namespace,
        period: Duration.minutes(1),
        statistic: 'Sum',
        dimensions: {
          TableName: tableName,
        },
      });
    }

    return {
      read: metric('ConsumedReadCapacityUnits', 'Consumed (Read)'),
      write: metric('ConsumedWriteCapacityUnits', 'Consumed (Write)'),
    };
  }
}
