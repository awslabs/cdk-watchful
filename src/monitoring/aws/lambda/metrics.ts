import { Metric } from '@aws-cdk/aws-cloudwatch';
import * as cdk from '@aws-cdk/core';

export class LambdaMetricFactory {
  static readonly Namespace = 'AWS/Lambda';

  metricInvocations(functionName: string) {
    return new Metric({
      metricName: 'Invocations',
      namespace: LambdaMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        FunctionName: functionName,
      },
    });
  }

  metricDuration(functionName: string) {
    function metric(statistic: string) {
      return new Metric({
        metricName: 'Duration',
        label: statistic,
        namespace: LambdaMetricFactory.Namespace,
        period: cdk.Duration.minutes(5),
        statistic,
        dimensions: {
          FunctionName: functionName,
        },
      });
    }

    return {
      min: metric('Minimum'),
      avg: metric('Average'),
      p50: metric('p50'),
      p90: metric('p90'),
      p99: metric('p99'),
      max: metric('Maximum'),
    };
  }

  metricErrors(functionName: string) {
    return new Metric({
      metricName: 'Errors',
      namespace: LambdaMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        FunctionName: functionName,
      },
    });
  }

  metricThrottles(functionName: string) {
    return new Metric({
      metricName: 'Throttles',
      namespace: LambdaMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        FunctionName: functionName,
      },
    });
  }
}
