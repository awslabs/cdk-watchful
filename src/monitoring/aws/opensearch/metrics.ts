import { Aws, Duration } from 'aws-cdk-lib';
import { MathExpression, Metric, Statistic } from 'aws-cdk-lib/aws-cloudwatch';

const enum ClusterMetrics {
  ClusterStatusGreen = 'ClusterStatus.green',
  ClusterStatusYellow = 'ClusterStatus.yellow',
  ClusterStatusRed = 'ClusterStatus.red',
  ShardsActive = 'Shards.active',
  ShardsUnassigned = 'Shards.unassigned',
  ShardsDelayedUnassigned = 'Shards.delayedUnassigned',
  ShardsActivePrimary = 'Shards.activePrimary',
  ShardsInitializing = 'Shards.initializing',
  ShardsRelocating = 'Shards.relocating',
  Nodes = 'Nodes',
  SearchableDocuments = 'SearchableDocuments',
  DeletedDocuments = 'DeletedDocuments',
  CPUUtilization = 'CPUUtilization',
  FreeStorageSpace = 'FreeStorageSpace',
  ClusterUsedSpace = 'ClusterUsedSpace',
  ClusterIndexWritesBlocked = 'ClusterIndexWritesBlocked',
  JVMMemoryPressure = 'JVMMemoryPressure',
  AutomatedSnapshotFailure = 'AutomatedSnapshotFailure',
  KMSKeyError = 'KMSKeyError',
  KMSKeyInaccessible = 'KMSKeyInaccessible',
  InvalidHostHeaderRequests = 'InvalidHostHeaderRequests',
  OpenSearchRequests = 'OpenSearchRequests',
  HTTP2XX = '2xx',
  HTTP3XX = '3xx',
  HTTP4XX = '4xx',
  HTTP5XX = '5xx',
}
const enum MasterMetrics {
  MasterCPUUtilization = 'MasterCPUUtilization',
  MasterJVMMemoryPressure = 'MasterJVMMemoryPressure',
  MasterReachableFromNode = 'MasterReachableFromNode',
  MasterSysMemoryUtilization = 'MasterSysMemoryUtilization',
}

const enum InstanceMetrics {
  SearchLatency = 'SearchLatency',
  IndexingLatency = 'IndexingLatency',
  SearchRate = 'SearchRate',
  IndexingRate = 'IndexingRate',
  ThreadpoolBulkQueue = 'ThreadpoolBulkQueue',
  ThreadpoolWriteQueue = 'ThreadpoolWriteQueue',
  ThreadpoolSearchQueue = 'ThreadpoolSearchQueue',
  ThreadpoolIndexQueue = 'ThreadpoolIndexQueue',
  ThreadpoolForceMergeQueue = 'ThreadpoolForce_mergeQueue',
  ThreadpoolBulkThreads = 'ThreadpoolBulkThreads',
  ThreadpoolWriteThreads = 'ThreadpoolWriteThreads',
  ThreadpoolSearchThreads = 'ThreadpoolSearchThreads',
  ThreadpoolIndexThreads = 'ThreadpoolIndexThreads',
  ThreadpoolForceMergeThreads = 'ThreadpoolForce_mergeThreads',
  ThreadpoolBulkRejected = 'ThreadpoolBulkRejected',
  ThreadpoolWriteRejected = 'ThreadpoolWriteRejected',
  ThreadpoolSearchRejected = 'ThreadpoolSearchRejected',
  ThreadpoolIndexRejected = 'ThreadpoolIndexRejected',
  ThreadpoolForceMergeRejected = 'ThreadpoolForce_mergeRejected',
  CoordinatingWriteRejected = 'CoordinatingWriteRejected',
  PrimaryWriteRejected = 'PrimaryWriteRejected',
  ReplicaWriteRejected = 'ReplicaWriteRejected',
}

const enum EbsMetrics {
  ReadLatency = 'ReadLatency',
  WriteLatency = 'WriteLatency',
  ReadThroughput = 'ReadThroughput',
  WriteThroughput = 'WriteThroughput',
  ReadIOPS = 'ReadIOPS',
  WriteIOPS = 'WriteIOPS'
}

const Namespace = 'AWS/ES';

/**
 * Metrics for OpenSearch
 */
export class OpenSearchMetricFactory {

  metricClusterStatus(domainName: string) {
    return {
      clusterStatusGreenMetric: this.metric(ClusterMetrics.ClusterStatusGreen, domainName).with({
        statistic: Statistic.MAXIMUM,
        period: Duration.minutes(1),
      }),
      clusterStatusRedMetric: this.metric(ClusterMetrics.ClusterStatusRed, domainName).with({
        statistic: Statistic.MAXIMUM,
        period: Duration.minutes(1),
      }),
      clusterStatusYellowMetric: this.metric(ClusterMetrics.ClusterStatusYellow, domainName).with({
        statistic: Statistic.MAXIMUM,
        period: Duration.minutes(1),
      }),
    };
  }

  metricFreeStorageSpace(domainName: string) {
    return this.metric(ClusterMetrics.FreeStorageSpace, domainName).with({ period: Duration.minutes(1) });
  }

  metricClusterIndexWritesBlocked(domainName: string) {
    return this.metric(ClusterMetrics.ClusterIndexWritesBlocked, domainName).with({ statistic: Statistic.MAXIMUM });
  }

  metricNodes(domainName: string) {
    return this.metric(ClusterMetrics.Nodes, domainName).with({ statistic: Statistic.MAXIMUM, period: Duration.hours(1) });
  }

  metricAutomatedSnapshotFailure(domainName: string) {
    return this.metric(ClusterMetrics.AutomatedSnapshotFailure, domainName).with({ statistic: Statistic.MAXIMUM, period: Duration.minutes(1) });
  }

  metricCpuUtilization(domainName: string) {
    return this.metric(ClusterMetrics.CPUUtilization, domainName).with({ statistic: Statistic.AVERAGE, period: Duration.minutes(15) });
  }

  metricJVMMemoryPressure(domainName: string) {
    return this.metric(ClusterMetrics.JVMMemoryPressure, domainName).with({ statistic: Statistic.MAXIMUM });
  }

  metricKms(domainName: string) {
    return {
      kmsKeyErrorMetric: this.metric(ClusterMetrics.KMSKeyError, domainName).with(
        {
          statistic: Statistic.MAXIMUM,
          period: Duration.minutes(1),
        }),
      kmsKeyInaccessibleMetric: this.metric(ClusterMetrics.KMSKeyInaccessible, domainName).with(
        {
          statistic: Statistic.MAXIMUM,
          period: Duration.minutes(1),
        }),
    };
  }

  metricShards(domainName: string) {
    return {
      shardsActive: this.metric(ClusterMetrics.ShardsActive, domainName).with({
        statistic: Statistic.SUM,
        period: Duration.minutes(1),
      }),
      shardsUnassigned: this.metric(ClusterMetrics.ShardsUnassigned, domainName).with({
        statistic: Statistic.SUM,
        period: Duration.minutes(1),
      }),
      shardsDelayedUnassigned: this.metric(ClusterMetrics.ShardsDelayedUnassigned, domainName).with({
        statistic: Statistic.SUM,
        period: Duration.minutes(1),
      }),
      shardsActivePrimary: this.metric(ClusterMetrics.ShardsActivePrimary, domainName).with({
        statistic: Statistic.SUM,
        period: Duration.minutes(1),
      }),
      shardsInitializing: this.metric(ClusterMetrics.ShardsInitializing, domainName).with({
        statistic: Statistic.SUM,
        period: Duration.minutes(1),
      }),
      shardsRelocating: this.metric(ClusterMetrics.ShardsRelocating, domainName).with({
        statistic: Statistic.SUM,
        period: Duration.minutes(1),
      }),
    };
  }

  metricHttp5xxPercentage(domainName: string) {
    const http5XXMetric = this.metricResponseCodes(domainName)['5xxMetric'];
    const openSearchRequestsMetric = this.metricRequests(domainName).openSearchRequests;

    return new MathExpression({
      expression: '100*(http5XXMetric/openSearchRequestsMetric)',
      period: Duration.minutes(1),
      usingMetrics: { http5XXMetric, openSearchRequestsMetric },
      label: 'HTTP5XXPercentage',
    });
  }

  metricThreadpoolQueues(domainName: string) {
    return {
      threadpoolBulkQueue: this.metric(InstanceMetrics.ThreadpoolBulkQueue, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
      threadpoolWriteQueue: this.metric(InstanceMetrics.ThreadpoolWriteQueue, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
      threadpoolSearchQueue: this.metric(InstanceMetrics.ThreadpoolSearchQueue, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
      threadpoolIndexQueue: this.metric(InstanceMetrics.ThreadpoolIndexQueue, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
      threadpoolForceMergeQueue: this.metric(InstanceMetrics.ThreadpoolForceMergeQueue, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
    };
  }

  metricThreadpoolThreads(domainName: string) {
    return {
      threadpoolBulkThreads: this.metric(InstanceMetrics.ThreadpoolBulkThreads, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
      threadpoolWriteThreads: this.metric(InstanceMetrics.ThreadpoolWriteThreads, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
      threadpoolSearchThreads: this.metric(InstanceMetrics.ThreadpoolSearchThreads, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
      threadpoolIndexThreads: this.metric(InstanceMetrics.ThreadpoolIndexThreads, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
      threadpoolForceMergeThreads: this.metric(InstanceMetrics.ThreadpoolForceMergeThreads, domainName).with( {
        statistic: Statistic.AVERAGE,
        period: Duration.minutes(1),
      }),
    };
  }

  metricThreadpoolRejected(domainName: string) {
    return {
      threadpoolBulkRejected: this.metric(InstanceMetrics.ThreadpoolBulkRejected, domainName).with( {
        period: Duration.minutes(1),
      }),
      threadpoolWriteRejected: this.metric(InstanceMetrics.ThreadpoolWriteRejected, domainName).with( {
        period: Duration.minutes(1),
      }),
      threadpoolSearchRejected: this.metric(InstanceMetrics.ThreadpoolSearchRejected, domainName).with( {
        period: Duration.minutes(1),
      }),
      threadpoolIndexRejected: this.metric(InstanceMetrics.ThreadpoolIndexRejected, domainName).with( {
        period: Duration.minutes(1),
      }),
      threadpoolForceMergeRejected: this.metric(InstanceMetrics.ThreadpoolForceMergeRejected, domainName).with( {
        period: Duration.minutes(1),
      }),
    };
  }

  metricWriteRejections(domainName: string) {
    return {
      CoordinatingWriteRejected: this.metric(InstanceMetrics.CoordinatingWriteRejected, domainName),
      PrimaryWriteRejected: this.metric(InstanceMetrics.PrimaryWriteRejected, domainName),
      ReplicaWriteRejected: this.metric(InstanceMetrics.ReplicaWriteRejected, domainName),
    };
  }

  metricRequests(domainName: string) {
    return {
      invalidHostHeaderRequests: this.metric(ClusterMetrics.InvalidHostHeaderRequests, domainName).with({
        period: Duration.minutes(1),
      }),
      openSearchRequests: this.metric(ClusterMetrics.OpenSearchRequests, domainName).with({
        period: Duration.minutes(1),
      }),

    };
  }

  metricStorage(domainName: string) {
    return {
      freeStorageSpaceMetric: this.metric(ClusterMetrics.FreeStorageSpace, domainName).with({
        statistic: Statistic.MINIMUM,
        period: Duration.minutes(1),
      }),
      clusterUsedSpaceMetric: this.metric(ClusterMetrics.ClusterUsedSpace, domainName).with({ statistic: Statistic.AVERAGE }),
    };
  }

  metricDocuments(domainName: string) {
    return {
      searchableDocumentsMetric: this.metric(ClusterMetrics.SearchableDocuments, domainName).with({ statistic: Statistic.AVERAGE }),
      deletedDocumentsMetric: this.metric(ClusterMetrics.DeletedDocuments, domainName).with({ statistic: Statistic.AVERAGE }),
    };
  }

  metricLatency(domainName: string) {
    return {
      indexingLatencyMetric: this.metric(InstanceMetrics.IndexingLatency, domainName).with({ statistic: Statistic.AVERAGE }),
      searchLatencyMetric: this.metric(InstanceMetrics.SearchLatency, domainName).with({ statistic: Statistic.AVERAGE }),
    };
  }

  metricRate(domainName: string) {
    return {
      indexingRateMetric: this.metric(InstanceMetrics.IndexingRate, domainName).with({ statistic: Statistic.AVERAGE }),
      searchRateMetric: this.metric(InstanceMetrics.SearchRate, domainName).with({ statistic: Statistic.AVERAGE }),
    };
  }

  metricClusterEbsLatency(domainName: string) {
    return {
      readLatencyMetric: this.metric(EbsMetrics.ReadLatency, domainName).with({ statistic: Statistic.AVERAGE }),
      writeLatencyMetric: this.metric(EbsMetrics.WriteLatency, domainName).with({ statistic: Statistic.AVERAGE }),
    };
  }
  metricClusterEbsThroughput(domainName: string) {
    return {
      readThroughputMetric: this.metric(EbsMetrics.ReadThroughput, domainName).with({ statistic: Statistic.AVERAGE }),
      writeThroughputMetric: this.metric(EbsMetrics.WriteThroughput, domainName).with({ statistic: Statistic.AVERAGE }),
    };
  }
  metricClusterEbsIOPS(domainName: string) {
    return {
      readIOPSMetric: this.metric(EbsMetrics.ReadIOPS, domainName).with({ statistic: Statistic.AVERAGE }),
      writeIOPSMetric: this.metric(EbsMetrics.WriteIOPS, domainName).with({ statistic: Statistic.AVERAGE }),
    };
  }

  metricResponseCodes(domainName: string) {
    return {
      '2xxMetric': this.metric(ClusterMetrics.HTTP2XX, domainName).with({ period: Duration.minutes(1) }),
      '3xxMetric': this.metric(ClusterMetrics.HTTP3XX, domainName).with({ period: Duration.minutes(1) }),
      '4xxMetric': this.metric(ClusterMetrics.HTTP4XX, domainName).with({ period: Duration.minutes(1) }),
      '5xxMetric': this.metric(ClusterMetrics.HTTP5XX, domainName).with({ period: Duration.minutes(1) }),
    };
  }

  metricMasterCpuUtilization(domainName: string) {
    return this.metric(MasterMetrics.MasterCPUUtilization, domainName).with({ statistic: Statistic.MAXIMUM, period: Duration.minutes(1) });
  }

  metricMasterJVMMemoryPressure(domainName: string) {
    return this.metric(MasterMetrics.MasterJVMMemoryPressure, domainName).with({ statistic: Statistic.MAXIMUM, period: Duration.minutes(1) });
  }

  metricMasterSysMemoryUtilization(domainName: string) {
    return this.metric(MasterMetrics.MasterSysMemoryUtilization, domainName).with({ statistic: Statistic.MAXIMUM, period: Duration.minutes(1) });
  }

  metricMasterReachableFromNode(domainName: string) {
    return this.metric(MasterMetrics.MasterReachableFromNode, domainName).with({ statistic: Statistic.MINIMUM, period: Duration.hours(1) });
  }

  protected metric(metric: ClusterMetrics | MasterMetrics | InstanceMetrics | EbsMetrics, domainName: string) {
    return new Metric({
      metricName: metric,
      namespace: Namespace,
      period: Duration.minutes(5),
      statistic: Statistic.SUM,
      dimensionsMap: {
        DomainName: domainName,
        ClientId: Aws.ACCOUNT_ID,
      },
    });
  }
}
