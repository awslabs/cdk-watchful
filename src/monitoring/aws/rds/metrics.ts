import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as cdk from '@aws-cdk/core';

/**
 * Metrics for RDS Aurora.
 */
export class RdsAuroraMetricFactory {
  static readonly Namespace = 'AWS/RDS';

  metricDmlThroughput(clusterIdentifier: string) {
    const dbInsertThroughputMetric = new cloudwatch.Metric({
      metricName: 'InsertThroughput',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
    const dbUpdateThroughputMetric = new cloudwatch.Metric({
      metricName: 'UpdateThroughput',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
    const dbSelectThroughputMetric = new cloudwatch.Metric({
      metricName: 'SelectThroughput',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
    const dbDeleteThroughputMetric = new cloudwatch.Metric({
      metricName: 'DeleteThroughput',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
    return {
      dbInsertThroughputMetric,
      dbUpdateThroughputMetric,
      dbSelectThroughputMetric,
      dbDeleteThroughputMetric,
    };
  }

  metricBufferCacheHitRatio(clusterIdentifier: string) {
    return new cloudwatch.Metric({
      metricName: 'BufferCacheHitRatio',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Average',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
  }

  metricDbConnections(clusterIdentifier: string) {
    return new cloudwatch.Metric({
      metricName: 'DatabaseConnections',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Average',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
  }

  metricReplicaLag(clusterIdentifier: string) {
    return new cloudwatch.Metric({
      metricName: 'AuroraReplicaLag',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Average',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
  }

  metricCpuUtilization(clusterIdentifier: string) {
    return new cloudwatch.Metric({
      metricName: 'CPUUtilization',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Average',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
  }
}
