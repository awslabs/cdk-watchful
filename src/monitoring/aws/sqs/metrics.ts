import { Metric, Statistic } from '@aws-cdk/aws-cloudwatch';
import { Duration } from '@aws-cdk/core';

const enum Metrics {
  ApproximateNumberOfMessagesVisible = 'ApproximateNumberOfMessagesVisible',
  NumberOfMessagesSent = 'NumberOfMessagesSent',
  ApproximateAgeOfOldestMessage = 'ApproximateAgeOfOldestMessage',
  SentMessageSize = 'SentMessageSize'
}

const Namespace = 'AWS/SQS';

export class SqsMetricFactory {
  metricApproximateVisibleMessages(topicName: string) {
    return this.metric(Metrics.ApproximateNumberOfMessagesVisible, topicName).with({ statistic: Statistic.MAXIMUM });
  }

  metricIncomingMessages(topicName: string) {
    return this.metric(Metrics.NumberOfMessagesSent, topicName).with({ statistic: Statistic.SUM });
  }

  metricAgeOfOldestMessageInSeconds(topicName: string) {
    return this.metric(Metrics.ApproximateAgeOfOldestMessage, topicName).with({ statistic: Statistic.MAXIMUM });
  }

  metricAverageMessageSizeInBytes(topicName: string) {
    return this.metric(Metrics.SentMessageSize, topicName).with({ statistic: Statistic.AVERAGE });
  }

  protected metric(metric: Metrics, topicName: string) {
    return new Metric({
      metricName: metric,
      namespace: Namespace,
      period: Duration.minutes(5),
      dimensions: {
        TopicName: topicName,
      },
    });
  }
}
