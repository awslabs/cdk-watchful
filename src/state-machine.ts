import { ComparisonOperator, GraphWidget } from '@aws-cdk/aws-cloudwatch';
import { StateMachine } from '@aws-cdk/aws-stepfunctions';
import { Construct, Duration } from '@aws-cdk/core';
import { IWatchful } from './api';
import { StateMachineMetricFactory } from './monitoring/aws/state-machine/metrics';

export interface WatchStateMachineOptions {
  /**
   * Alarm when execution failures reach this threshold over 1 minute.
   *
   * @default 1 any execution failure will trigger the alarm
   */
  readonly metricFailedThreshold?: number;
}

export interface WatchStateMachineProps extends WatchStateMachineOptions {
  readonly title: string;
  readonly watchful: IWatchful;
  readonly stateMachine: StateMachine;
}

export class WatchStateMachine extends Construct {
  private readonly title: string;
  private readonly watchful: IWatchful;
  private readonly stateMachine: StateMachine;
  private readonly metricFailedThreshold: number;

  private readonly metrics: StateMachineMetricFactory;

  constructor(scope: Construct, id: string, props: WatchStateMachineProps) {
    super(scope, id);

    this.title = props.title;
    this.watchful = props.watchful;
    this.stateMachine = props.stateMachine;

    this.metricFailedThreshold = props.metricFailedThreshold ?? 1;

    this.metrics = new StateMachineMetricFactory();

    this.createLinks();
    this.createExecutionMetrics();
  }

  private createExecutionMetrics() {
    const execMetrics = this.metrics.metricExecutions(this.stateMachine.stateMachineArn);
    const { failed } = execMetrics;
    const failureAlarm = failed.createAlarm(this, 'ExecutionFailures', {
      alarmDescription: `at ${this.metricFailedThreshold}`,
      threshold: this.metricFailedThreshold,
      period: Duration.minutes(5),
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      evaluationPeriods: 1,
      statistic: 'sum',
    });

    this.watchful.addAlarm(failureAlarm);

    this.watchful.addWidgets(new GraphWidget({
      title: 'Overall Execution/min',
      width: 12,
      stacked: false,
      left: Object.values(execMetrics),
      leftAnnotations: [{ value: this.metricFailedThreshold, color: '#ff0000', label: 'Execution Failure Alarm' }],
    }));
  }

  private createLinks() {
    this.watchful.addSection(this.title, {
      links: [{
        title: 'Amazon State Machine',
        url: `https://console.aws.amazon.com/states/home?region=${this.stateMachine.stack.region}#/statemachines/view/${this.stateMachine.stateMachineArn}`,
      }],
    });
  }
}