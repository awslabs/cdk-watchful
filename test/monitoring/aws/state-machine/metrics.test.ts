import { StateMachineMetricFactory } from '../../../../src/monitoring/aws/state-machine/metrics';

const dummyStateMachineArn = 'dummy-state-machine-arn';

test('snapshot test: metricExecutions', () => {
  // GIVEN
  const unitToTest = new StateMachineMetricFactory();

  // WHEN
  const metric = unitToTest.metricExecutions(dummyStateMachineArn);

  // THEN
  expect(metric.failed.metricName).toStrictEqual('ExecutionsFailed');
  Object.values(metric).forEach(eachMetric => {
    expect(eachMetric.dimensions).toStrictEqual({ StateMachineArn: dummyStateMachineArn });
  });
  expect(metric).toMatchSnapshot();
});