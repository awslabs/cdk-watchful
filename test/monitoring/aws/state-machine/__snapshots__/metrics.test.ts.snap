// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: metricExecutions 1`] = `
Object {
  "aborted": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "StateMachineArn": "dummy-state-machine-arn",
    },
    "label": "Aborted Executions",
    "metricName": "ExecutionsAborted",
    "namespace": "AWS/States",
    "period": Duration {
      "amount": 1,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Sum",
    "unit": undefined,
  },
  "failed": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "StateMachineArn": "dummy-state-machine-arn",
    },
    "label": "Failed Executions",
    "metricName": "ExecutionsFailed",
    "namespace": "AWS/States",
    "period": Duration {
      "amount": 1,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Sum",
    "unit": undefined,
  },
  "succeeded": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "StateMachineArn": "dummy-state-machine-arn",
    },
    "label": "Executions Succeeded",
    "metricName": "ExecutionsSucceeded",
    "namespace": "AWS/States",
    "period": Duration {
      "amount": 1,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Sum",
    "unit": undefined,
  },
  "throttled": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "StateMachineArn": "dummy-state-machine-arn",
    },
    "label": "Executions Throttled",
    "metricName": "ExecutionThrottled",
    "namespace": "AWS/States",
    "period": Duration {
      "amount": 1,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Sum",
    "unit": undefined,
  },
  "timedOut": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "StateMachineArn": "dummy-state-machine-arn",
    },
    "label": "Executions TimedOut",
    "metricName": "ExecutionsTimedOut",
    "namespace": "AWS/States",
    "period": Duration {
      "amount": 1,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Sum",
    "unit": undefined,
  },
  "total": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "StateMachineArn": "dummy-state-machine-arn",
    },
    "label": "Total",
    "metricName": "ExecutionsStarted",
    "namespace": "AWS/States",
    "period": Duration {
      "amount": 1,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Sum",
    "unit": undefined,
  },
}
`;
